test2code/src/main/kotlin/com/epam/drill/plugins/test2code/coverage/ProbeArrayProvider.kt

```diff
@@ -0,0 +1,186 @@
+
+    // TODO EPMDJ-8256 When application is async we must use this implementation «com.alibaba.ttl.TransmittableThreadLocal»
+    private val requestThreadLocal = ThreadLocal<ExecData?>()
+
+    private val job = ProbeWorker.launch(start = CoroutineStart.LAZY) {
+        while (true) {
+            delay(2000L)
+            realtimeHandler(collectProbes())
+        }
+    }
+
+    private var globalExecData = ExecData()
+    private val stubProbes = StubAgentProbes()
+    private val probeMetaContainer = ConcurrentHashMap<ClassId, ProbeDescriptor>()
+    private val execDataPool: DataPool<SessionTestKey, ExecData> = ConcurrentDataPool()
+
+    override fun invoke(
+        id: Long,
+        num: Int,
+        name: String,
+        probeCount: Int,
+    ): AgentProbes = getSessionClassProbes(id)
+        ?: getGlobalClassProbes(id)
+        ?: stubProbes
+
+    override fun addProbeDescriptor(descriptor: ProbeDescriptor) {
+        probeMetaContainer[descriptor.id] = descriptor
+        globalExecData.getOrPut(descriptor.id) {
+            descriptor.toExecDatum()
+        }
+    }
+
+    override fun ExecData.fillExecData(
+        sessionId: String,
+        testId: String,
+        testName: String
+    ) {
+        probeMetaContainer.values.forEach { descriptor ->
+            this.putIfAbsent(
+                descriptor.id, ExecDatum(
+                    id = descriptor.id,
+                    name = descriptor.name,
+                    probes = AgentProbes(descriptor.probeCount),
+                    sessionId = sessionId,
+                    testName = testName,
+                    testId = testId
+                )
+            )
+        }
+    }
+
+    override fun startRecording(sessionId: String, testId: String, testName: String) {
+        val data = execDataPool.getOrPut(SessionTestKey(sessionId, testId to testName)) {
+            ExecData().apply { fillExecData(sessionId, testId, testName) }
+        }
+        requestThreadLocal.set(data)
+        logger.trace { "Test recording started (sessionId = $sessionId, testId=$testId, threadId=${Thread.currentThread().id})." }
+    }
+
+    override fun stopRecording(sessionId: String, testId: String, testName: String) {
+        val data = requestThreadLocal.get()
+        if (data != null) {
+            execDataPool.release(SessionTestKey(sessionId, testId to testName), data)
+            requestThreadLocal.remove()
+        }
+        logger.trace { "Test recording stopped (sessionId = $sessionId, testId=$testId, threadId=${Thread.currentThread().id})." }
+    }
+
+    override fun collectProbes(): Sequence<ExecDatum> {
+        releaseGlobalExecData()
+        return execDataPool.pollReleased().flatMap { it.values }.filter { it.probes.hasPositive() }
+    }
+
+    private fun releaseGlobalExecData() {
+        globalExecData.values.filter { datum ->
+            datum.probes.hasPositive()
+        }.map { datum ->
+            val probesToSend = datum.probes.values.copyOf()
```

Comment:
I know this art of code migrated from previous implementation, but it is contrived and makes little sense.
TBD - refactoring

Reply:
TODO - refactoring